### DISCUSSION TITLE
What IAM actions do lambda functions need?

### DISCUSSION THREAD

**Discussion by zfLQ2qx2** [cloud-custodian/cloud-custodian#7321] [General]
Good evening,

For an EC2 policy that is going to consume cloudwatch events, check attributes of an EC2 instance, and submit a message to SQS for c7n_mailer, what are the minimal set of IAM permissions I would need for the lambda role? 

I am guessing:
logs:CreateLogGroup
logs:CreateLogStream
logs:PutLogEvents
ec2:DescribeInstances
ec2:DescribeTags
sqs:SendMessage

I'm not sure about cloudwatch or what other ec2 actions I might need. Maybe ec2:Describe*

**Comment by thisisshi** [comment]
~You will also need sts:AssumeRole~, but your current list looks like a good start

each filter/action is also annotated with the appropriate iam permissions like so: https://github.com/capitalone/cloud-custodian/blob/master/c7n/resources/ec2.py#L80

additionally, you'll need the following permissions to deploy custodian: 

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "events:DescribeRule",
                "events:EnableRule",
                "events:ListTargetsByRule",
                "events:PutRule",
                "events:PutTargets",
                "iam:PassRole",
                "lambda:AddPermission",
                "lambda:CreateAlias",
                "lambda:CreateFunction",
                "lambda:GetAlias",
                "lambda:GetFunction",
                "lambda:ListTags",
                "lambda:UpdateAlias",
                "lambda:UpdateFunctionCode",
                "lambda:UpdateFunctionConfiguration"
            ],
            "Resource": "*"
        }
    ]
}
```
source: https://github.com/capitalone/cloud-custodian/issues/1693#issuecomment-376284377

Edit: assumerole isn't needed, was looking at a different set of perms regarding LambdaMode

**Comment by kapilt** [comment]
you don't need sts assume role, unless your passing --assume on the cli. if your not doing that you'll need to explicitly set a role on your lambda policy.

wrt to the permissions for a given policy to execute in lambda, its highly dependent on what filters and actions your taking.

**Comment by zfLQ2qx2** [comment]
@thisisshi @kapilt  Thank you for the information!
